src/lem/var_map.rs

```diff
@@ -0,0 +1,46 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use super::Var;
+
+/// `VarMap` is a wrapper around a `HashMap` whose keys are `Var`s. It's meant
+/// to be more ergonomic under the following assumptions for LEM:
+///
+/// 1. A LEM must be in SSA, so we don't want to bind some information to the
+/// same `Var` twice;
+/// 2. A LEM must always define variables before using them, so we don't expect
+/// to need some piece of information from a variable that hasn't been defined.
+#[derive(Clone)]
+pub struct VarMap<V>(HashMap<Var, V>);
+
+impl<V> VarMap<V> {
+    /// Creates an empty `VarMap`
+    #[inline]
+    pub(crate) fn new() -> VarMap<V> {
+        VarMap(HashMap::default())
+    }
+
+    /// Inserts new data into a `VarMap`
+    ///
+    /// Panics if some data already exists for the key
+    pub(crate) fn insert(&mut self, var: Var, v: V) {
+        match self.0.entry(var) {
+            Entry::Vacant(vacant_entry) => {
+                vacant_entry.insert(v);
+            }
+            Entry::Occupied(o) => panic!(
+                "Variable {} has previously been defined. LEMs are supposed to be SSA.",
```

Comment:
Nit: the very strong requirement is panic inducing, which makes me quite uncomfortable, if only because to put decent UX around this (to e.g. show the user *where*. they violated ssa) we’ll now have to rely on panic recovery. But probably outside the scope of this PR.